wasm.rkt

```diff
@@ -60,14 +63,92 @@
 ;; Delta
 
 (define-metafunction WASM-eval
-  delta : op v ... -> v
+  delta : op v ... -> v ∨ trap
+  ;;; BINOPS
   [(delta add (const c_0) (const c_1)) (const ,(+ (term c_0) (term c_1)))]
+  [(delta sub (const c_0) (const c_1)) (const ,(- (term c_0) (term c_1)))]
+  [(delta mul (const c_0) (const c_1)) (const ,(* (term c_0) (term c_1)))]
+  [(delta div (const c_0) (const c_1)) ,(if (= (term c_1) 0)
+                                            (term trap)
+                                            (term (const ,(/ (term c_0) (term c_1)))))]
+  [(delta rem (const c_0) (const c_1)) ,(if (= (term c_1) 0)
+                                            (term trap)
+                                            (term (const ,(remainder (term c_0) (term c_1)))))]
+  [(delta and (const c_0) (const c_1)) (const ,(bitwise-and (term c_0) (term c_1)))]
+  [(delta or (const c_0) (const c_1)) (const ,(bitwise-ior (term c_0) (term c_1)))]
+  [(delta xor (const c_0) (const c_1)) (const ,(bitwise-xor (term c_0) (term c_1)))]
+  ; Note that the shift and rotate ops only make sense in the context of a fixed bit width.
+  ; Here, we assume that all operands are 32 bit.
+  [(delta shl (const c_0) (const c_1))
+   (const ,(bitwise-and #xFFFFFFFF
+                        (arithmetic-shift (term c_0) (term c_1))))]
+  [(delta shr (const c_0) (const c_1))
+   (const ,(arithmetic-shift (term c_0) (- (term c_1))))]
+  [(delta rotl (const c_0) (const c_1))
+   (const ,(if (< (term c_1) 0)
+               (term (rotr (const c_0) (const ,(- (term c_1)))))
+               (bitwise-ior (bitwise-and #xFFFFFFFF
+                                         (arithmetic-shift (term c_0) (remainder (term c_1) 32)))
+                            (arithmetic-shift (term c_0)
+                                              (- (remainder (term c_1) 32) 32)))))]
+  [(delta rotr (const c_0) (const c_1))
+   (const ,(if (< (term c_1) 0)
+               (term (rotl (const c_0) (const ,(- (term c_1)))))
+               (bitwise-ior (arithmetic-shift (term c_0) (- (remainder (term c_1) 32)))
+                            (bitwise-and #xFFFFFFFF
+                                         (arithmetic-shift (term c_0)
+                                                           (- 32 (remainder (term c_1) 32)))))))]
```

Comment:
Good catch, there is some extra complexity here because there is some ambiguity as to the sign of the result, since we aren't specifying the signedness of the input.

Reply:
I think this should be addressed in a follow-up PR, since there are a lot of implications in specifying the sign, such as in the comparison ops

Reply:
I'm good with addressing it later - could we add a comment saying that all the ops only work for signed 32 bit positive values (and not signed negative values)?

Reply:
Done